Design/Instrument-2.0/design-notes.md

```diff
@@ -0,0 +1,89 @@
+- `DetectorInfo`, a class that stores (and provides access to) basic parameters of detectors, such as L2, twoTheta, position, masking flags, monitor flags, ....
+- `SpectrumInfo`, a class that stores (and provides access to) basic parameters of spectra, such as L2, twoTheta, position, masking flags, monitor flags, .... The values hare are derived from those on `DetectorInfo`. This class reflects the common case in Mantid that a spectrum can represent more than one detector, e.g., due to grouping of detectors or after summing spectra.
+
+
+## Notes
+
+### Spectrum-number and detector ID
+
+Mantid currently works with `SpectrumNumber` and `DetectorID` in many places.
+As discussed in detail in this [design proposal](https://github.com/mantidproject/documents/blob/spectrum_number_and_workspace_index_abstraction/Design/spectrum_number_and_workspace_index_abstraction.md), these concepts should not be spread accross our codebase.
+Apart from the arguments presented there, early tests with the [prototype](https://github.com/DMSC-Instrument-Data/instrument-prototype) showed that using `DetectorID` as an index for accessing detector data is inefficient, since the resulting `std::map` lookups slow things down significantly.
+
+Therefore, neither `SpectrumNumber` nor `DetectorID` should be part of the core mechanics of the instrument.
+Translation to a plain index, e.g., for accessing a `std::vector` should happen as close to the user interface as possible.
+Mechanism for this translation would probably be part of the workspace interface, but separate from, e.g., `SpectrumInfo`.
+The remainder of this document is thus not concerned with `SpectrumNumber` or `DetectorID`.
+
+### Naming
+
+All class names in this document are preliminary and just for the sake of discussion.
+Better names are needed in several cases.
+
+
+## Design
+
+### Structure
+
+Algorithms that work with workspaces need to access information about spectra.
+Therefore:
+
+- `SpectrumInfo` should be a member of workspace.
+- `SpectrumInfo` owns `DetectorInfo`.
+- `DetectorInfo` owns `InstrumentTree`.
+
+If no detectors are grouped, there is a 1:1 mapping between spectrum and detector.
+`SpectrumInfo` and `DetectorInfo` should thus share data, to avoid waste of memory and CPU resources.
+This can be facilitated by:
+
+- `GeometryDataArray` is used to store vectors of data, such as L2, twoTheta, position, masking flags, monitor flags, ....
+- `SpectrumInfo` and `DetectorInfo` have a field `std::shared_ptr<GeometryDataArray>`. By default both point to the same instance. If detectors are grouped, `SpectrumInfo` allocates its own `GeometryDataArray`.
+
+### Interface
+
+The interface to `SpectrumInfo` and `DetectorInfo` would be almost identical.
+The most typical examples for public methods include:
+
+```cpp
+bool isMonitor(size_t index);
+bool isMasked(size_t index);
+double L2(size_t index);
+double twoTheta(size_t index);
+V3D position(size_t index);
```

Comment:
@quantumsteve thanks for pointing [this](https://gitlab.kitware.com/vtk/vtk/blob/master/Documentation/Doxygen/ArrayDispatch-VTK-7-1.md) out. The SOA approach seems to make a lot of sense, and given that VTK have been able to take generic compile-time approaches to handling these is encouraging. However, I have several issues with taking the SOA approach when it comes to V3D.

* We have situations where we want to take a polymorphic approach to `Components`, of which `Detectors` are a subclass. Currently `Detectors` are "discovered" and registered at runtime in to the `InstrumentTree` as a AOS (see [here](https://github.com/DMSC-Instrument-Data/instrument-prototype/blob/master/cow_instrument/InstrumentTree.h#L37)). To do the equivalent using the SOA approach, and maintain the ability to return an individual `Detector` such as is done [here](https://github.com/DMSC-Instrument-Data/instrument-prototype/blob/master/cow_instrument/InstrumentTree.h#L26) we'd need some way to slice through our SOA and create the `Detectors` on the fly. 
* Certain geometric operations are still the responsibility of `Components` such as positioning and rotating. Adding two `V3D`s together seems like a sensible operation and provides a nice encapsulation. The same would be true of `Quat` for the equivalent rotation matrix. Furthermore the new `InstrumentTree` relies on knowing and walking sub-trees for write operations. How does one handle this in a SOA configuration. Currently each detector has an index, so it is possible to walk the tree and index into the SOA.

What are peoples thoughts?